Ported interleaving support to System.IO.Packaging

edwardneal opened the pull request:
Contributes to (and appears to fix) #51929.
Also contributes to (and appears to fix) dotnet/wpf#3546.

This was previously part of PR #78374, (originally by @KevinCathcart 's work) which was closed due to poor test coverage in System.IO.Packaging. I've added a set of tests for my changes, (although there's some overlap with normal functionality) and corrected a few deviations.

To try to keep the internal state of ZipPackage manageable, I've introduced a behaviour change: users can no longer try to create an atomic part (i.e. a part backed by a single-ZipArchiveEntry) with a URI which belongs to an interleaved part; this means that trying to pass a URI of `/[Content_Types].xml/[0].piece` to `Package.CreatePart` will fail. I don't think this'll actually have an impact though - it's a violation of the OPC spec, and would have resulted in inconsistent behaviour anyway.

The OPC spec also states that the piece number should be a single digit. ZipPackage is more forgiving in this PR, for two reasons:
1. Compatibility with the .NET Framework (which has the same behaviour)
2. The XPS Document Writer (and presumably, other applications) generates packages which have many pieces. I tested with one of the files in [this comment](https://github.com/dotnet/wpf/issues/3546#issuecomment-824713800) and found parts with thousands of pieces

I've made a few changes to KevinCathcart's original code, but most of them are stylistic. The only key differences are in the InterleavedZipPackagePartStream - it would try to seek the underlying streams, and calling SetLength didn't always work because ZipArchiveEntry had a stream left open.

edwardneal committed:
Added interleaving support to System.IO.Packaging

This is essentially KevinCathcart's work, with a test suite and changes required to ensure that these tests pass.

edwardneal committed:
Removed reference to NuGet.Frameworks

Also removed other unused "using" statements.

huoyaoyuan commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,535 @@
+        private static void SeekUnderlyingPieceStream(Stream pieceStream, long byteCount)
+        {
+            // need to add a test for all of this
+            const int BufferSize = 4096;
+            long remainingBytes = byteCount;
+            byte[] readBuffer = new byte[BufferSize];
```
Use array from pool, and use ReadExactly to avoid looping.

edwardneal replied:
I'll change it to use a pooled array, thanks. I don't want to use ReadExactly though, due to memory usage. Using ReadExactly requires that I allocate an array of size `byteCount`, and this could be very large - if an uncompressed part piece won't fit in memory then I think it's better for the method to take a long time than to crash.

edwardneal replied:
Looking more broadly at the method: I want the calling of this to be a corner case. It's trying to seek to a specific position in a stream that doesn't support seeking; that's never going to be efficient.

If a ZipArchiveEntry has a compression method of Stored, the underlying stream will be a SubReadStream. This doesn't support seeking right now, for some reason - I've got a PR in mind to add that. DeflateStreams are by definition, much harder though.

One way to support ad-hoc seeks into a DeflateStream might be to cache X position mappings from the uncompressed to the compressed content as we read, at intervals of `UncompressedLength / X`. A seek could set the position to the best position mapping and read forwards from there, or there could be a `TrySeek` method on a decompressing stream; writing would invalidate relevant parts of the cache. The same concept might also apply to other decompression methods.

This was the point where I decided to keep the PR on track... but this method's got a high performance penalty; I'd like to see it lowered, but it requires upstream changes of some kind.

huoyaoyuan commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackage.cs:
```diff
@@ -459,6 +534,168 @@ private static CompressionOption GetCompressionOptionFromZipFileInfo()
+                else
+                {
+                    //Check for incomplete sequence.
+                    if (string.CompareOrdinal(pieces[i].NormalizedPrefixName, normalizedPrefixNameForCurrentSequence) != 0)
```
Use `!=` for ordinal equality comparison

edwardneal replied:
Yes, agreed.

edwardneal commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,535 @@
+        /// </param>
+        /// <param name="count">How many bytes requested.</param>
+        /// <returns>How many bytes were written into buffer.</returns>
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            CheckClosed();
+
+            // Check arguments.
+            PackagingUtilities.VerifyStreamReadArgs(this, buffer, offset, count);
```
That makes perfect sense, thanks - I'll make those changes today

edwardneal replied:
I've made the changes, but the caveat here is that .NET Standard 2.0 didn't include span-based read and writes. I've had to thread the Spans through to the part piece streams, which isn't possible for .NET Standard and is mandatory for decent performance in .NET Core 2.2+.

My workaround has been to temporarily rent byte buffers, copy the spans to/from them and perform the reads. This'll effectively double memory usage, but I don't think it'll happen in the real world: .NET Framework already has interleaving support, and we'd need to go back to an unsupported version of .NET Core for this to be a problem.

huoyaoyuan replied:
You can follow System.IO.Pipelines which also supports netstandard2.0:

https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs

Define a `ReadInternal`/`ReadCore` method that uses span. Include the `Read(Span)` override only in .NET Core

edwardneal replied:
I see what you mean about the external declaration of Read(Span) and Write(Span) now, and have removed it from the .NET Standard 2.0 build

edwardneal commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,535 @@
+                    SeekUnderlyingPieceStream(pieceStream, pieceStreamRelativeOffset);
+                }
+
+                while (totalBytesRead < count)
```
The loop is needed because when Read returns zero, we need to move to the next part piece and continue reading.

edwardneal committed:
Changes following code review

Removed use of string.CompareOrdinal when processing part pieces.
Replaced byte allocations with ArrayPool Rent/Returns.
Made InterleavedZipPackagePartStream use Spans instead of byte arrays. This has increased memory usage when .NET Framework calls Read(byte[], int, int) and Write(byte[], int, int).

edwardneal committed:
Adjusted the compilation conditions for Read

Split Read and ReadCore apart, so that a new Read(Span<byte>) method doesn't appear on the .NET Standard 2.0 path. Also adjusted the conditions from "NETCOREAPP2_1_OR_GREATER" to "!NETFRAMEWORK && !NETSTANDARD2_0", to align with the rest of .NET.

edwardneal committed:
Moved the ArrayPool.Return calls into a finally block, for safety purposes

I can't see a simple way to do this in the Write() path - we rent and return across some loop iterations

KevinCathcart commented:
I should probably mention that my work here was mostly just porting back the code from .NET Framework. While essentially all of that code got contributed under MIT between the WPF repository, and this repository, portions were ifdefed out, and some was modified in the initial commits, so I did need to reference the actual .NET Framework sources in a few spots  (via the referencesource site, not the MIT reference source repo, which lacked needed libraries), to verify what the code originally looked like. I feel like anything taken form there is likely too small to matter, and probably was authorized to be relicensed anyway. Presumably the changes in the initial commits vs framework were likely not for licensing reasons, but stemmed from a desire to have the initially checked in code able to actually compile. 

Of course there was some original work to get things to fit on top of ZipStream instead of MS.Internal.IO.Zip, and any original contributions of mine there are covered under MIT, and my CLA with the .NET Foundation.

wstaelens commented:
@huoyaoyuan can you review please?

huoyaoyuan commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,572 @@
+        /// <inheritdoc/>
+        protected override void Dispose(bool disposing)
+        {
+
```
Nit: extra new lines (and another few cases)

huoyaoyuan commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,572 @@
+            {
+                do
+                {
+                    bytesRead = pieceStream.Read(readBuffer, 0, bytesRead);
```
The buffer length calculation looks incorrect here. It reads at most BufferSize.
Pseudo code should be:

```
remainingBytes = arg1;
byte[] buffer = Rent(BufferSize);
do
{
    // note that buffer can be larger than requested size and we can utilize it
    bytesRead = pieceStream.Read(buffer, 0, Min(buffer.Length, remainingBytes);
    remainingBytes -= bytesRead;
} while (remainingBytes > 0 && bytesRead > 0);
```

huoyaoyuan commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackage.cs:
```diff
@@ -976,5 +1318,216 @@ private static void ThrowIfXmlAttributeMissing(string attributeName, string? att
+                                entry?.Delete();
+                            }
+
+                            /*foreach (ZipArchiveEntry entry in _zipArchive.Entries)
```
What is the commented code for?

huoyaoyuan commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.PieceDirectory.cs:
```diff
@@ -0,0 +1,456 @@
+            /// </summary>
+            internal PieceDirectory(List<ZipPackagePartPiece> sortedPieceInfoList, ZipStreamManager zipStreamManager, FileAccess access)
+            {
+                if (sortedPieceInfoList == null)
+                    throw new ArgumentNullException(nameof(sortedPieceInfoList));
```
For non-public methods, we usually do not check for null and let the caller ensure it instead.

huoyaoyuan commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.PieceDirectory.cs:
```diff
@@ -0,0 +1,456 @@
+                Debug.Assert(sortedPieceInfoList.Count > 0);
+
+                // Initialize the first piece.
+                _pieceStreamInfoList = [
+                    new PieceStreamInfo(
+                    zipStreamManager.Open(sortedPieceInfoList[0].ZipArchiveEntry, access),
+                    0 /*startOffset*/)
+                ];
```
Style: `[` on new line, indent for parameters of the constructor call, use `startOffset: 0` for naming parameter

huoyaoyuan commented:
I'm not area owner or experienced about System.IO.Packaging. I can only provide review for general practices. Please wait for the area owners.

edwardneal commented:
Thanks - I agree with your comments. The Read logic makes sense, not quite sure what the thought process was there! I'll take a look this evening and resolve them.

No worries with the PR timing. It's a little larger than I'd normally want to submit, it's a feature implementation rather than a bugfix and I imagine the preview release has taken some of the teams' bandwidth, so I appreciate it'll take longer to review.

danmoseley commented:
@richlander  the code on referencesource - is it all under MIT like the reference source repo is?

edwardneal committed:
Changes following second round of code review

* Corrections to comments and spacing.
* Switched to the use of an EndOfStreamException in the correct situation.
* Adjusted the SeekUnderlyingPieceStream method to slightly tweak the mechanism and to make use of a readBuffer which might be larger than BufferSize.
* Small refactor of part piece deletion.

edwardneal commented:
The post-code review changes are completed now, thanks @huoyaoyuan. I'm not convinced that the buffer length calculation in `SeekUnderlyingPieceStream` was incorrect, but it was definitely unclear and it wouldn't always use the entire buffer (if ArrayPool.Rent returned an array which was larger than expected.) I've changed it to make it a little clearer.

wstaelens commented:
> The post-code review changes are completed now, thanks @huoyaoyuan. I'm not convinced that the buffer length calculation in `SeekUnderlyingPieceStream` was incorrect, but it was definitely unclear and it wouldn't always use the entire buffer (if ArrayPool.Rent returned an array which was larger than expected.) I've changed it to make it a little clearer.

https://github.com/dotnet/runtime/blob/e9218122ae717a393198be04c611061a14f6ee88/src/libraries/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs#L81

shouldn't this (and others in the class) be https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception.throwifnull?view=net-7.0 or is it because of compatibility with .net standard?

edwardneal commented:
That's correct, yes - the code needs to compile under .NET Standard 2.0, which excludes that particular helper method.

A pattern match would definitely look a bit better, but it wouldn't avoid a cast. This:
```csharp
if (!(partUri is ValidatedPartUri validatedUri))
    Console.WriteLine("Hello, World!");
```
will actually compile down to:
```csharp
ValidatedPartUri validatedUri = partUri as ValidatedPartUri;
if (validatedUri != null)
    Console.WriteLine("Hello, World!");
```

Eliminating the second cast could be helpful though; I'll make the changes tomorrow morning to incorporate any intermediary changes.

edwardneal committed:
Eliminating duplicate casts, minor comment cleanup

wstaelens commented:
@edwardneal what is the status of this? Will it be in .net 9 (and maybe backported?)

edwardneal commented:
I've not got a status update to share unfortunately. I'd be glad to see this in .NET 9, but it's currently waiting on code review from the System.IO area owners.

A .NET 6.0/7.0/8.0 project could still reference the updated System.IO.Packaging NuGet package; as a result, the older project should use the .NET Standard 2.0 build within the package and be able to use the fixed functionality. This'll hopefully do what you're looking for, but if it doesn't and the .NET team approve the PR for backport, I'm happy enough to help where needed.

wstaelens commented:
@dotnet/area-system-io @jeffhandley @ericstj ?

huoyaoyuan commented:
Features won't be backported. The OOB package can provide support for lower versions.

wstaelens commented:
> I've not got a status update to share unfortunately. I'd be glad to see this in .NET 9, but it's currently waiting on code review from the System.IO area owners.
> 
> A .NET 6.0/7.0/8.0 project could still reference the updated System.IO.Packaging NuGet package; as a result, the older project should use the .NET Standard 2.0 build within the package and be able to use the fixed functionality. This'll hopefully do what you're looking for, but if it doesn't and the .NET team approve the PR for backport, I'm happy enough to help where needed.

@richlander @stephentoub @terrajobst any news or updates? I see .net 9 preview 5 but no news on this...

carlossanlop committed:
Merge branch 'main' into issue-51929

ViktorHofer commented on src/libraries/System.IO.Packaging/src/System.IO.Packaging.csproj:
```diff
@@ -39,10 +41,13 @@
     <Compile Include="System\IO\Packaging\PackUriHelper.PackUriScheme.cs" />
     <Compile Include="$(CommonPath)System\Obsoletions.cs" Link="Common\System\Obsoletions.cs" />
+
+    <InternalsVisibleTo Include="System.IO.Packaging.Tests" Key="00240000048000009400000006020000002400005253413100040000010001004b86c4cb78549b34bab61a3b1800e23bfeb5b3ec390074041536a7e3cbd97f5f04cf0f857155a8928eaa29ebfd11cfbbad3ba70efea7bda3226c6a8d370a4cd303f714486b6ebc225985a638471e6ef571cc92a4613c00b8fa65d61ccee0cbe5f36330c9a01f4183559f1bef24cc2917c6d913e3a541333a1d05d9bed22b38cb" />
```
If possible, it would be good to not depend on `InternalsVisibleTo` as that hinders the linker from removing internal types which could cause size concerns or regressions from the last known good state.

Alternatives to IVT that we usually recommend are source sharing or reflection.

edwardneal replied:
Thanks - I see what you mean. I'm pretty sure the tests just depend directly on `ZipPackagePartPiece.TryParse`, so reflection sounds like the best way to go here. I'll change this shortly

edwardneal committed:
Replaced InternalsVisibleTo

This was only used to reference ZipPackagePartPiece.TryParse and .Index, so switched the tests to use reflection.

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs:
```diff
@@ -0,0 +1,214 @@
+        private const string PartPieceFileExtension = ".piece";
+        private const string PartPieceLastExtension = "].last";
+
+        #region Internal Methods
```
nit: Can you please remove regions from new code please?

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs:
```diff
@@ -0,0 +1,214 @@
+            //  prefix_name "/" "[" 1*digit "]" [".last"] ".piece"
+            // Work backwards from the end of the full path, extracting and checking this metadata in stages.
+            // Stage 1: extract the file extension of ".piece".
+            int resultPosition = path.LastIndexOf(PartPieceFileExtension, StringComparison.OrdinalIgnoreCase);
```
Does the spec mention anything about case sensitivity? Can the extension be `.PIECE`?

edwardneal replied:
Yes. ECMA-376 part 2, section 6.2.2.3 says:

> Equivalence of part names shall be determined by ASCII case-insensitive matching. Such matching 
compares a sequence of code points as if all ASCII code points in the range 0x41–0x5A (A–Z) were mapped 
to the corresponding code points in the range 0x61–0x7A (a–z). See Reference [1]

Section 7.2.5.2 says:

> If a logical item is a piece, its name shall have suffixes of the following syntax:
> `SuffixName = "/" "[" PieceNumber "]" [".last"] ".piece"`
> `PieceNumber = "0" | NonZeroDigit [1*Digit]`
> `Digit = "0" | NonZeroDigit`
> `NonZeroDigit = "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" `
> The prefix of a logical item name is the result of removing a suffix, if any, from the logical item name.
> Equivalence of prefixes, and of suffixes shall be determined by ASCII case-insensitive matching. Logical 
names shall be equivalent if their prefixes and suffixes are equivalent. A physical package shall not contain 
equivalent logical item names. 

We validate case-insensitivity in the `ValidPartPiecesAreParsable` test.

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs:
```diff
@@ -0,0 +1,214 @@
+            //  prefix_name "/" "[" 1*digit "]" [".last"] ".piece"
+            // Work backwards from the end of the full path, extracting and checking this metadata in stages.
+            // Stage 1: extract the file extension of ".piece".
+            int resultPosition = path.LastIndexOf(PartPieceFileExtension, StringComparison.OrdinalIgnoreCase);
```
`PartPieceFileExtension` does not seem to be used anywhere else but here. I can directly be passed here.
```suggestion
            int resultPosition = path.LastIndexOf(".piece", StringComparison.OrdinalIgnoreCase);
```

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs:
```diff
@@ -0,0 +1,214 @@
+    /// </summary>
+    internal sealed class ZipPackagePartPiece : IComparable<ZipPackagePartPiece>
+    {
+        private const string PartPieceFileExtension = ".piece";
```
See comment below.
```suggestion
```

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs:
```diff
@@ -0,0 +1,214 @@
+                int digitStart;
+
+                // Iterate backwards, character by character, until we find a non-digit character.
+                for (digitStart = searchPosition; digitStart > 1 && char.IsDigit(path[digitStart - 1]); digitStart--)
+                    ;
```
```suggestion
                for (digitStart = searchPosition; digitStart > 1 && char.IsDigit(path[digitStart - 1]); digitStart--);
```

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs:
```diff
@@ -0,0 +1,214 @@
+        #region Internal Constructors
+        internal ZipPackagePartPiece(ZipArchiveEntry zipArchiveEntry, PackUriHelper.ValidatedPartUri? partUri, string prefixName, int pieceNumber, bool isLastPiece)
+        {
+            if (zipArchiveEntry == null)
+                throw new ArgumentNullException(nameof(zipArchiveEntry));
+            if (prefixName == null)
+                throw new ArgumentNullException(nameof(prefixName));
+            if (pieceNumber < 0)
+                throw new ArgumentOutOfRangeException(nameof(pieceNumber));
```
Do we care if `prefixName` is an empty string? In case we need its exception to be `ArgumentNullException.ThrowIfNull` instead.

```suggestion
            ArgumentNullException.ThrowIfNull(zipArchiveEntry, nameof(zipArchiveEntry));
            ArgumentException.ThrowIfNullOrEmpty(prefixName, nameof(prefixName));
            ArgumentOutOfRangeException.ThrowIfNegative(pieceNumber, nameof(pieceNumber));
```

edwardneal replied:
`prefixName` can't be empty - the prefix is the parent directory, and part pieces in the root directory are invalid. The `InvalidPartPiecesAreNotParsable` test validates that these are blocked.

I've switched everything to use the exception helper methods (although System.IO.Packaging is built for .NET Standard, so I've left the originals in situ for that.)

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,572 @@
+                            totalBytesRead,
+                            buffer.Length - totalBytesRead);
+
+                        tempInputBuffer.AsSpan(totalBytesRead, numBytesRead).CopyTo(buffer.Slice(totalBytesRead, numBytesRead));
```
@huoyaoyuan do these changes match your suggestions?

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,572 @@
+        {
+            CheckClosed();
+
+            // Check stream capabilities. (Normally, CanSeek will be false only
+            // when the stream is closed.)
```
I don't know about this comment. Do we have tests that verify the behavior when handling unseekable streams as the input? For example, a network stream?

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,572 @@
+            CheckClosed();
+
+            // Check argument and stream capabilities.
+            if (newLength < 0)
+                throw new ArgumentOutOfRangeException(nameof(newLength));
```
```suggestion
            ArgumentOutOfRangeException.ThrowIfNegative(newLength, nameof(newLength));
```

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,572 @@
+        }
+
+        // High-level object to access the collection of pieces by offset and pieceNumber.
+        private readonly PieceDirectory _dir;
```
nit: Do you mind moving the private members to the beginning of the class?

carlossanlop commented on src/libraries/System.IO.Packaging/tests/PartPieceTests.cs:
```diff
@@ -0,0 +1,606 @@
+            zipArchive.Dispose();
+        }
+    }
+}
```
Can you please add at least one test for an unseekable stream? We need to confirm this code works for, say, a network stream as an input.

You can continue using a MemoryStream as an input, but it needs to be wrapped with [a WrappedStream](https://github.com/dotnet/runtime/blob/cae3aec80d32316026d85228e12a94dd4887d57f/src/libraries/Common/tests/System/IO/WrappedStream.cs). There are many examples in System.Formats.Tar, like [this one](https://github.com/dotnet/runtime/blob/cae3aec80d32316026d85228e12a94dd4887d57f/src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.Tests.cs#L544).

edwardneal replied:
Would you mind clarifying please? The part pieces are built on `ZipArchiveEntry` objects, and the streams returned by `ZipArchiveEntry` are always non-seekable on .NET Core (even if the stream the Package's `ZipArchive` was opened from is seekable.)

edwardneal replied:
I understand now, thanks. I'd expect different behaviour though, generated by `ZipArchive`.

| Package.Open `packageAccess` | Resultant ZipArchiveMode | Non-seekable stream behaviour |
| - | - | - |
| `FileAccess.Read` | `ZipArchiveMode.Read` | Stream copied to (seekable) MemoryStream |
| `FileAccess.Write` | `ZipArchiveMode.Create` | No impact - can only be used to create Packages |
| `FileAccess.ReadWrite` (default) | `ZipArchiveMode.Update` | ZipArchive constructor will throw |

I expect this'll throw within [the ZipArchive constructor](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L120) before it even reaches the code under test in Package. I'll add tests which verify these behaviours.

edwardneal replied:
I see your point now, thanks - there's a behavioural difference between `ZipArchive` on .NET Core and Framework. We're now testing both behaviours, bringing us to three tests:
1. (.NET Core) Can create a package and add a part when the package is backed by an unseekable stream.
2. (.NET Framework) Cannot create a package when the package is backed by an unseekable stream.
3. Cannot open a package for modification when the package is backed by an unseekable stream.

Once the tests complete in CI, I'll mark this as resolved.

carlossanlop commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs:
```diff
@@ -0,0 +1,572 @@
+        /// is a descriptor for the 1st piece.
+        /// </para>
+        /// </remarks>
+        public override bool CanSeek
+        {
+            get
+            {
+                return _closed ? false : _dir.GetStream(0).CanSeek;
+            }
+        }
```
```suggestion
        public override bool CanSeek =>_closed ? false : _dir.GetStream(0).CanSeek;
```

carlossanlop reviewed (commented):
Thank you so much for your change, @edwardneal . I left some comments for you to consider. The one I care the most about is making sure we are able to handle unseekable streams properly. I left a comment in the test file with a suggestion.

edwardneal committed:
Code review

* Using exception helpers for .NET
* Removing #regions
* Fixing nullability in references to ZipPackagePartPiece.PieceDescriptors
* Spacing/spelling/grammar changes to comments

edwardneal committed:
Added tests for non-seekable streams

edwardneal committed:
Swapped conditional compilation to newer form

Replaced references to NETFRAMEWORK and NETSTANDARD2_0 with NET and !NET as appropriate.

edwardneal commented on src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs:
```diff
@@ -0,0 +1,215 @@
+        /// </remarks>
+        internal static bool TryParse(ZipArchiveEntry zipArchiveEntry, [NotNullWhen(true)] out ZipPackagePartPiece? partPiece)
+        {
+#if !NETFRAMEWORK && !NETSTANDARD2_0
```
Thanks @kasperk81 - I've just updated these references.

carlossanlop commented on src/libraries/System.IO.Packaging/tests/Tests.NetFramework.cs:
```diff
@@ -0,0 +1,28 @@
+            using (MemoryStream ms = new MemoryStream())
+            using (WrappedStream ws = new WrappedStream(ms, true, true, false))
+            {
+                Assert.Throws<ArgumentException>(() => Package.Open(ws, FileMode.Create, FileAccess.Write));
```
I see the .NET Framework version of Package.Open goes through [here](https://referencesource.microsoft.com/#WindowsBase/Base/System/IO/Packaging/Package.cs,263) which in turns calls [this](https://referencesource.microsoft.com/#WindowsBase/Base/System/IO/Packaging/Package.cs,1050) with  `streaming` set to `false`.

But then I'm not sure who's throwing `ArgumentException`. Do you have a callstack?

edwardneal replied:
It's the `streaming` parameter which generates it. It gets passed into [ZipPackage](https://referencesource.microsoft.com/#WindowsBase/Base/System/IO/Packaging/ZipPackage.cs,437), then into [ZipArchive.OpenOnStream](https://referencesource.microsoft.com/#WindowsBase/Base/MS/Internal/IO/Zip/ZipArchive.cs,126), then [ValidateModeAccessShareStreaming](https://referencesource.microsoft.com/#WindowsBase/Base/MS/Internal/IO/Zip/ZipArchive.cs,524), then [ValidateModeAccessStreamStreamingCombinations](https://referencesource.microsoft.com/#WindowsBase/Base/MS/Internal/IO/Zip/ZipArchive.cs,569), where an exception is thrown if `streaming` and `CanSeek` are both false.

The full call stack is:
```
ZipArchive.ValidateModeAccessStreamStreamingCombinations(Stream stream, FileAccess access, Boolean streaming)
ZipArchive.ValidateModeAccessShareStreaming(Stream stream, FileMode mode, FileAccess access, FileShare share, Boolean streaming)
ZipArchive.OpenOnStream(Stream stream, FileMode mode, FileAccess access, Boolean streaming)
ZipPackage.ctor(Stream s, FileMode mode, FileAccess access, Boolean streaming)
Package.Open(Stream stream, FileMode packageMode, FileAccess packageAccess, Boolean streaming)
Tests.Cannot_Create_Package_On_Unseekable_Stream() line 22
```

carlossanlop commented on src/libraries/System.IO.Packaging/tests/Tests.cs:
```diff
@@ -3961,6 +3961,23 @@ public void Roundtrip_Compression_Option(CompressionOption createdCompressionOpt
+
+                using (WrappedStream ws = new WrappedStream(ms, true, true, false))
+                {
+                    Assert.Throws<ArgumentException>(() => Package.Open(ws, FileMode.Open, FileAccess.ReadWrite));
```
Looks good.

The Package.Open method is [here](https://source.dot.net/#System.IO.Packaging/System/IO/Packaging/Package.cs,886), and it then sets the `ZipArchiveMode` to `Update` [here](https://source.dot.net/#System.IO.Packaging/System/IO/Packaging/ZipPackage.cs,326), which is expected to throw [here](https://source.dot.net/#System.IO.Compression/System/IO/Compression/ZipArchive.cs,151).

carlossanlop commented on src/libraries/System.IO.Packaging/tests/Tests.Net.cs:
```diff
@@ -0,0 +1,39 @@
+            using (MemoryStream ms = new MemoryStream())
+            using (WrappedStream ws = new WrappedStream(ms, true, true, false))
+            {
+                Package package = Package.Open(ws, FileMode.Create, FileAccess.Write);
```
This looks good. It's the one I was expecting to work because we support Create mode with unseekable stream [here](https://source.dot.net/#System.IO.Compression/System/IO/Compression/ZipArchive.cs,160).

richlander commented:
When is the P7 snap @carlossanlop? We should try to get this one in.

carlossanlop reviewed (approved):
This is ready to merge. P7 snap is next week. This will be part of it.

Thanks so much @edwardneal for working on this, and also thank you for your patience.

Edit: I was just waiting for the CI re-runs to pass, there was a ton of noise.

danmoseley commented:
For the folks tracking this -- it would be great if you could get the preview 7 build as soon as its out (or any daily build that has this in) and validate that you're getting good results.

edwardneal commented:
Sorry for the delay here, but I've successfully tested with a known-interleaved XPS document from @wstaelens original issue dotnet/wpf#3546 - the file in `02` from [this comment](https://github.com/dotnet/wpf/issues/3546#issuecomment-824713800).

Sample code:
```csharp
using System.IO.Packaging;

const string FileName = "02.xps";
const string InterleavedPart = "/FixedDocumentSequence.fdseq";

using var package = Package.Open(FileName);
var parts = package.GetParts();
var namedPart = package.GetPart(new Uri(InterleavedPart, UriKind.Relative));

Console.WriteLine($"Loaded file {Path.GetFileName(FileName)}");
Console.WriteLine($"Parts: {parts.Count()}");
Console.WriteLine($"Found named part \"{InterleavedPart}\": {namedPart is not null}");

if (namedPart is not null)
{
    using var partStream = namedPart.GetStream(FileMode.Open, FileAccess.Read);
    using var reader = new StreamReader(partStream);

    Console.WriteLine("Contents of named part:");
    Console.WriteLine(reader.ReadToEnd());
}
```

Thanks to everyone for your reviews, and for pushing the merge through @carlossanlop.

wstaelens commented:
Currently on holiday without pc access and limited internet access.
I could have tested the same code as in my original ticket as @edwardneal did to see if that worked.

(Sent already to my colleagues but did not hear from them yet)

danmoseley commented:
> Sorry for the delay here, but I've successfully tested with a known-interleaved XPS document from @wstaelens original issue https://github.com/dotnet/wpf/issues/3546 - the file in 02 from https://github.com/dotnet/wpf/issues/3546#issuecomment-824713800.

I want to add a woohoo here, I moved to ASP.NET since this issue was opened but really happy to see it happen, and kudos again for the hard work and contributions of everyone here.

cc @pchaurasia14
